### ISSUE TITLE
adding activated policies to the audit log

### ISSUE THREAD

**Issue by cornelinux** [privacyidea/privacyidea#874]
Split from #829 

I think this is also a good idea! If we do not want to change the audit table schema, we could also add a DEBUG logging output that sums up the activated policies for each request in a first step.

On the technical side, I think we have to differentiate between pre- and postpolicies:

I noticed that the prepolicies all return True in any case, but the return value doesn't seem to be processed. Couldn't we use the return value to indicate whether the policy was activated? The prepolicy decorator could then construct a list of activated policies in the request context (e.g. g.activated_prepolicies), i.e. a list of all policy functions that returned True.
For postpolicies, it seems to be a bit harder because they return the new response. Maybe the postpolicy decorator could check whether the response was modified by the postpolicy? (e.g. if new_response != old_response or new_response.data != old_data: ...).
Or we make this explicit and add a function announce_policy_activation() or something like that. We would just need to insert calls to that function at the right places.

**Comment by cornelinux** [comment]
In a first step we could also write this to the privacyIDEA log.

**Comment by cornelinux** [comment]
I just noticed a problem: Decorators like ``check_serial`` https://github.com/privacyidea/privacyidea/blob/master/privacyidea/api/lib/postpolicy.py#L233
use the function ``get_action_values`` to determine, if something should, happen or an authentication should fail. 

But ``get_action_values`` only returns the value and not the policy object or the policy name, that provided this value. I.e. the decorator knows a configured value, but does not know, which policy provided this value. 

So we do not know the policy name and can not add it to the audit object in the decorator!

**Comment by cornelinux** [comment]
The *prepolicy decorator* change the request data while the *postpolicy policy* decorator change the response data.

@fredreichbier  I have an simple example of what you suggested in #829 to do this centrally in the postpolicy wrapper. It works nicely with a few lines of code. But I do not have access to the ``g`` object. Thus I can not easily pass this to the audit log.

As a POC I can write it to the log file.

**Comment by cornelinux** [comment]
The problem is that in this scenario only the _actions_ are logged to the logfile.

E.g.:

> [2018-02-12 16:12:16,415][24112][140582688536320][INFO][privacyidea.api.lib.prepolicy:124] The API request to http://localhost:5000/validate/check was modified by post policy: mangle

So we know, that the call was mangled, but not by *which* policy definition. (Could be more policies with the action ``mangle``.)

**Comment by cornelinux** [comment]
It might be more sensible to add the information within the policy.
Although this might mean one additional line per policy decorater, we can add more usefull information...
...like here:
https://github.com/privacyidea/privacyidea/blob/master/privacyidea/api/lib/prepolicy.py#L824

like:

    g.audit_object.log({"policies": ....add_policy_name_...    )

**Comment by fredreichbier** [comment]
I agree that having the policy names (instead of the action names only) would be useful information, even at the cost of one additional line per policy function. And I also think that adding a function to  the audit object would be sensible. As a first step, maybe the base function in ``Audit`` could also just log to the debug log?
But does the SQL audit table provide enough space to store the activated policies? Or should we add an additional column?

**Comment by cornelinux** [comment]
I stumpled upon this yesterday. In a support request we were wondering, why the system responded, as it did. Which policies were used during authentication?
The list of triggered policies would help a lot, when trying to understand, why the authentication succeeded or failed.

So I looked at it again.
To my understanding policies are defined in
* https://github.com/privacyidea/privacyidea/blob/master/privacyidea/api/lib/postpolicy.py
* https://github.com/privacyidea/privacyidea/blob/master/privacyidea/api/lib/prepolicy.py
* https://github.com/privacyidea/privacyidea/blob/master/privacyidea/lib/policydecorators.py

Thankfully, in all three modules we have ``g``. And thus we also have ``g.audit_object``.

We tried the wrapper for **post policies** and **pre policies** https://github.com/privacyidea/privacyidea/blob/874/triggered_policy_in_audit_log/privacyidea/api/lib/postpolicy.py#L108 and https://github.com/privacyidea/privacyidea/blob/874/triggered_policy_in_audit_log/privacyidea/api/lib/prepolicy.py#L121.

But the wrapper only has the method name of the policy - not the policy name itself.
Maybe we can find a wrapper solution for also adding the name. Or we do it in each policy. We have ~ 30 pre policies and 12 post policies.

For the [policydecorator.py](https://github.com/privacyidea/privacyidea/blob/master/privacyidea/lib/policydecorators.py) I have no problem with adding an update to ``g.audit_object`` in each policy method.
After all there are only **10** policy methods - easy!

In regards to the problem with ``policy_object.get_action_values``, which does not reveal the name of the policy, that provided the value, we can enhance this to also (somehow) return the policy name.

**Comment by fredreichbier** [comment]
I've pushed an idea to [874/log-activated-policies](https://github.com/privacyidea/privacyidea/compare/874/log-activated-policies). I had the idea of using the loglevel ``logging.INFO - 1``, because that allows users to get all INFO messages *as well as* the activated policies, but *not* the DEBUG messages (because they are a lot). But I'm not sure if that's a good idea. :-)

**Comment by cornelinux** [comment]
I added a Policy-Object-Method which will return *actionvalues* and their corresponding *policy name* while keeping backward compat. This way we could replace the methods step by step or add the audit log step by step.

The question still is: Should we enhance the audit log table?
I think so.
This way the admin could also search for events, where a certain policy was triggered!

**Comment by fredreichbier** [comment]
I think ``get_action_values_with_name`` is a good idea. I think we would then add something like
```python
action_dict = policy_object.get_action_values_with_name(...)
# and only if the policies takes effect:
audit_object.log({"policies": list(action_dict.keys())})
```
to the policy functions?

Regarding the audit log table: Yes, I think we should rather enhance it than try to squeeze the information in the currently existing columns. But this would mean we need to include the policy names in the signature. I think @plettich mentioned we should improve the audit log signature approach too? Maybe we could do both changes in one step.

**Comment by plettich** [comment]
The thing is, in order to validate the signatures of each audit-log entry, we need the information with which data the signature was calculated. So if we add/remove data, we need some kind of versioning for the signature.

**Comment by cornelinux** [comment]
We could simply add a string column, that contains the names of all triggered policies.
If the list is too long, we could start truncating the names of the policies.

E.g. there are policies like:

* enroll_sha256_for_users
* disallow_totp_on_DMZ

to name a few, a string like "enroll_sha256_for_users,disallow_totp_on_DMZ" could easily get to long.

Instead of truncating only the end of the string (which could lead to a missing policy, if the last policy would be truncated completely), we could truncate *each* policy name like

"enroll_sha256_for_u,disallow_totp_on_"

This way we would be able to see all entries and also to filter for policies.

As an alternative we could save the database id of the policies (as a comma separated list like 1,10,12,34) and then reconstruct the names of the policy in the API response. 

@fredreichbier @plettich what do you think?

**Comment by fredreichbier** [comment]
I would prefer saving the (possibly truncated) policy names instead of policy IDs, because the policy IDs could change (during export/import, for example?) which would make the information useless. But policy names are hopefully descriptive and would provide some useful information even if the policies have changed in the meantime?

**Comment by cornelinux** [comment]
This is an interesting point, I did not think about!

However, I would like to add a link to the details view of the policies in the Audit UI, this link would break, if the policynames would be truncated. But maybe this is something we have to live with.

**Comment by plettich** [comment]
The filter would only work if the truncated name is known (or by educated guessing).
The truncation should be reproducible for the signature verification.
Should we provide a truncated name with each policy?
What is the limit on the column width?

**Comment by cornelinux** [comment]
The signature would of course run over the column which already contains the truncated information.
So no need to do any guessing in the code.
For the signature this would only be a textfield without any logic.

**Comment by cornelinux** [comment]
I added a truncation function in 3216ba1 to be used in the audit log 551180f so that policynames will be equally truncated. In the test I simply used German city names. Truncation works fine :-)

**Comment by cornelinux** [comment]
We will use the method 

~~~~python
audit_object.add_to_log({"policy": "{0},".format(policyname)})
~~~~
 to add policies to the audit log.

**Comment by cornelinux** [comment]
Closed by #1297
